### PR TITLE
[fix][speed] Better projections - correctness + speed

### PR DESCRIPTION
**Pull request by blefaudeux**
## What does this PR do?
- fixes the self attention optimization (project all in one go) not being triggered
- contiguous tensors are only needed for sparse, removing the contiguous cost when not needed
- fixing a simplifying a broken path in the factory / residual and normalization, related to the above: only normalize once if this is the same tensor, and propagate the same id down

With these changes/fixes, vanilla xformers is 5-10% faster than timm on a vanilla ViT (up from being 10% slower) as per the bench [from the other PR](https://github.com/facebookresearch/xformers/issues/108), and microGPT trains something decent in 20 minutes on a laptop (3080 / fp16)

## Before submitting

- [x] Did you have fun?
  - Make sure you had fun coding 🙃
- [x] Did you read the [contributor guideline](https://github.com/facebookresearch/fairscale/blob/master/CONTRIBUTING.md)?
- [ ] Was this discussed/approved via a Github issue? (no need for typos, doc improvements)
  - [ ] N/A
- [x] Did you make sure to update the docs?
  - [ ] N/A
- [ ] Did you write any new necessary tests?
  - [ ] N/A
- [ ] Did you update the [changelog](https://github.com/facebookresearch/fairscale/blob/master/CHANGELOG.md)? (if needed)
  - [ ] N/A

## PR review
Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

### CODE REVIEW

File: xformers/components/attention/_sputnik_sparse.py [hunk 1]
```diff
@@ -258,7 +258,7 @@ def matmul_with_mask(self, a, b):
         column_indices = self.column_indices
         out = _sddmm.apply(
             a,
-            b.transpose(-2, -1),
+            b.transpose(-2, -1).contiguous(),
```
Comment (blefaudeux) [inline xformers/components/attention/_sputnik_sparse.py:261]: contiguous is only needed here and down, we were forcing this all the time before. Note that with a good projection kernel this could come for free, could be nice

File: xformers/components/in_proj_container.py [hunk 2]
```diff
@@ -153,17 +153,14 @@ def forward(
...
 
                 if self.in_proj_bias is not None:
                     qkv += self.in_proj_bias
 
-                q, k, v = map(
-                    lambda x: x.contiguous(),
-                    qkv.split(self.out_features, dim=-1),
-                )
-                return q, k, v
+                qkv = qkv.split(self.out_features, -1)
```
Comment (blefaudeux) [inline xformers/components/in_proj_container.py:162]: this was actually slow (memcopy) and not needed for non-sparse attention

File: xformers/components/residual.py [hunk 3]
```diff
@@ -36,16 +27,23 @@ class LayerNormStyle(str, Enum):
...
+
+    .. warning: by convention, if multiple tensors are being passed in,
+        the first one is used for the residual path
+    """
 
     def __init__(self, layer: nn.Module):
         super().__init__()
         self.layer = layer
 
-    def forward(self, inputs: Union[torch.Tensor, List[torch.Tensor]], *args, **kwargs):
```
Comment (blefaudeux) [inline xformers/components/residual.py:45]: this was super error prone (@stephenroller spotted that long ago, I should have caught that then), since intputs and args can mix

File: xformers/components/residual.py [hunk 4]
```diff
@@ -62,11 +60,16 @@ def __init__(self, d_model: int, sublayer: nn.Module, use_triton: bool = True):
...
-        inputs = _to_tensor_list(inputs)
-
-        x_norm = [self.norm(x_) for x_ in inputs]
-        return self.sublayer(*x_norm, *args, **kwargs)
+    def forward(self, *args, **kwargs):
+        # Could be that the same tensor has been passed multiple times
+        # in that case we'll just normalize once
+        list_ids = [id(inp) for inp in args]
```
Comment (blefaudeux) [inline xformers/components/residual.py:66]: seems small, but here if the same tensor was actually passed multiple times (self attention), we would normalize 3 times and loose the same id(), which in turn means that in the attention layer we would not optimize for self-attention

File: xformers/factory/block_factory.py [hunk 5]
```diff
@@ -364,8 +364,8 @@ def forward(
         else:
             target_q, target_k, target_v = target, target, target
 
-        x = self.wrap_att([target_q, target_k, target_v], att_mask=decoder_att_mask)
-        x = self.wrap_cross([x, memory, memory], att_mask=encoder_att_mask)
+        x = self.wrap_att(target_q, target_k, target_v, att_mask=decoder_att_mask)
```
Comment (blefaudeux) [inline xformers/factory/block_factory.py:367]: related to the "input<>args" cleanup

**Comment by blefaudeux** [comment]
@fmassa @dianaml0 I think that the code in block_factory and residual is probably way to complicated and error prone, could be worth another pass if deemed important (this is orthogonal to the parts zoo approach). In any case with these changes a default xformers is competitive in terms of speed vs. Timm ViT

**Comment by blefaudeux** [comment]
Ping reviewers (and PR below), this significantly speeds up a code path used in the examples and by some folks around

**Comment by dianaml0** [comment]
> @fmassa @dianaml0 I think that the code in block_factory and residual is probably way to complicated and error prone, could be worth another pass if deemed important (this is orthogonal to the parts zoo approach). In any case with these changes a default xformers is competitive in terms of speed vs. Timm ViT

Seems like much of the complication was introduced for reversible layers to work? But agree probably a good idea to see if there's a cleaner way to do it.

**Review by dianaml0 — APPROVED** [review]
This is really great!! LGTM!

**Comment by blefaudeux** [comment]
> > @fmassa @dianaml0 I think that the code in block_factory and residual is probably way to complicated and error prone, could be worth another pass if deemed important (this is orthogonal to the parts zoo approach). In any case with these changes a default xformers is competitive in terms of speed vs. Timm ViT
> 
> 
> 
> Seems like much of the complication was introduced for reversible layers to work? But agree probably a good idea to see if there's a cleaner way to do it.

Yes for reversible, I agree, it makes the signal paths a lot more complex unfortunately:( no strong opinion on that, I don't think that this was benchmarked enough to have a definitive answer. Probably possible to have a better code and keep reversible though

File: xformers/components/attention/_sputnik_sparse.py (hunk 1 above)
Comment (stephenroller) [reply xformers/components/attention/_sputnik_sparse.py:261]: maybe you should leave a comment/breadcrumb trail saying that in the code?

Comment (blefaudeux) [reply xformers/components/attention/_sputnik_sparse.py:261]: will do, or maybe just file an issue, basically the projection in the beginning of MHA could be hardened a little for speed
